File: src/main/java/at/flauschigesalex/defaultLibrary/file/JsonManager.java
package at.flauschigesalex.defaultLibrary.file;

import lombok.Getter;
import org.bson.BsonDocument;
import org.bson.Document;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.File;
import java.net.http.HttpResponse;
import java.util.*;
import java.util.concurrent.atomic.AtomicBoolean;

@SuppressWarnings({"unused", "DataFlowIssue", "unchecked", "UnusedReturnValue"})
@Getter
public class JsonManager {

    public static JsonManager copy(final @NotNull JsonManager json) {
        return new JsonManager(json.content);
    }

    public static JsonManager copyOriginal(final @NotNull JsonManager json) {
        return new JsonManager(json.originalContent);
    }

    private final String originalContent;
    private String content;
    private FileHandler fileHandler;

    public JsonManager() {
        this("{}");
    }

    public JsonManager(final @NotNull Object content) {
        this(content.toString());
    }

    public JsonManager(final @NotNull Map<String, Object> map) {
        this();
        this.writeMany(map);
    }

    public JsonManager(final @NotNull Document document) {
        this(document.toJson());
    }

    public JsonManager(final @NotNull HttpResponse<String> response) {
        this(response.body());
    }

    public JsonManager(final @NotNull File file) {
        this(new FileHandler(file).readString());
    }

    public JsonManager(final @NotNull ResourceHandler resource) {
        this(resource.readString());
    }

    JsonManager(final @NotNull String content) {
        try {
            new JSONParser().parse(content);
        } catch (final Exception e) {
            throw new IllegalArgumentException("Failed to read json from string:\n"+content);
        }

        this.originalContent = content;
        this.content = content;
    }

    public Object asObject() {
        try {
            return new JSONParser().parse(getContent());
        } catch (ParseException ignore) {
        }
        return null;
    }
    public JSONObject asJsonObject() {
        try {
            return (JSONObject) asObject();
        } catch (Exception ignore) {
        }
        return null;
    }
    public JSONArray asJsonArray() {
        try {
            return (JSONArray) asObject();
        } catch (Exception ignore) {
        }
        return new JSONArray();
    }
    @Deprecated(forRemoval = true) public String asJsonString() {
        return content;
    }

    private Object get(final @NotNull JSONObject jsonObject, final @NotNull String path) {
        if (!jsonObject.containsKey(path))
            return null;
        return jsonObject.get(path);
    }

    public Object getObject(final @NotNull String path) {
        if (path.isEmpty())
            return asObject();

        JSONObject jsonObject = asJsonObject();
        if (jsonObject == null)
            return null;

        if (!path.contains(".")) {
            return get(jsonObject, path);
        }
        JSONObject current = jsonObject;
        String[] splitSourcePath = path.split("\\.");
        for (String splitSource : splitSourcePath) {
            Object object = get(current, splitSource);
            if (object == null)
                return null;

            if (path.endsWith(splitSource))
                return current.get(splitSource);
            try {
                current = (JSONObject) object;
            } catch (Exception ignore) {
                break;
            }
        }
        return null;
    }
    public JSONObject getJsonObject(final @NotNull String path) {
        if (path.isEmpty())
            return asJsonObject();

        try {
            return (JSONObject) getObject(path);
        } catch (Exception ignore) {
        }
        return null;
    }
    public JsonManager getJsonManager(final @NotNull String path) {
        try {
            return new JsonManager(getJsonObject(path));
        } catch (Exception ignore) {
        }
        return null;
    }

    public JSONArray getJsonArray(final @NotNull String path) {
        try {
            return (JSONArray) getObject(path);
        } catch (Exception ignore) {
        }
        return new JSONArray();
    }
    public ArrayList<Object> getObjectList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Object>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<String> getStringList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<String>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<Integer> getIntegerList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Integer>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<Long> getLongList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Long>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<Short> getShortList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Short>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<Double> getDoubleList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Double>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<Float> getFloatList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Float>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }
    public ArrayList<Boolean> getBooleanList(final @NotNull String path) {
        try {
            return new ArrayList<>((ArrayList<Boolean>) getObject(path));
        } catch (Exception ignore) {
        }
        return new ArrayList<>();
    }

    public String getString(final @NotNull String path) {
        Object object = getObject(path);
        if (object == null)
            return null;
        return object.toString();
    }
    public Integer getInteger(final @NotNull String path) {
        try {
            return Integer.parseInt(getString(path));
        } catch (Exception ignore) {
        }
        return null;
    }
    public Long getLong(final @NotNull String path) {
        try {
            return Long.parseLong(getString(path));
        } catch (Exception ignore) {
        }
        return null;
    }
    public Short getShort(final @NotNull String path) {
        try {
            return Short.parseShort(getString(path));
        } catch (Exception ignore) {
        }
        return null;
    }
    public Double getDouble(final @NotNull String path) {
        try {
            return Double.parseDouble(getString(path));
        } catch (Exception ignore) {
        }
        return null;
    }
    public Float getFloat(final @NotNull String path) {
        try {
            return Float.parseFloat(getString(path));
        } catch (Exception ignore) {
        }
        return null;
    }
    public Boolean getBoolean(final @NotNull String path) {
        try {
            return Boolean.parseBoolean(getString(path));
        } catch (Exception ignore) {
        }
        return null;
    }

    public boolean copyOf(final @NotNull String path, final @NotNull String newPath) {
        return this.copyOf(path, newPath, true);
    }
    public boolean copyOf(final @NotNull String path, final @NotNull String newPath, final boolean override) {
        if (!this.has(path) && !override)
            return false;
        if (this.has(newPath) && !override)
            return false;

        return this.write(newPath, this.getObject(path));
    }

    public boolean move(final @NotNull String path, final @NotNull String newPath) {
        return this.move(path, newPath, true);
    }
    public boolean move(final @NotNull String path, final @NotNull String newPath, final boolean override) {
        if (!this.has(path) && !override)
            return false;
        if (this.has(newPath) && !override)
            return false;

        if (!this.write(newPath, this.getObject(path)))
            return false;
        return this.remove(path);
    }

    public boolean removeMany(final @NotNull String... paths) {
        return this.removeMany(List.of(paths));
    }
    public boolean removeMany(final @NotNull Collection<String> paths) {
        final AtomicBoolean success = new AtomicBoolean(true);
        for (final String path : paths)
            if (!this.remove(path))
                success.set(false);

        return success.get();
    }
    public boolean remove(final @NotNull String path) {
        if (!this.has(path))
            return true;

        if (!path.contains(".")) {
            final var jsonObject = this.asJsonObject();
            jsonObject.remove(path);
            this.content = jsonObject.toString();

            this.checkRemoveEmpty();
            return true;
        }

        final var child = List.of(path.split("\\.")).getLast();
        final var parent = path.substring(0, path.length() - (child.length() + 1));

        final var jsonObject = this.getJsonObject(parent);
        jsonObject.remove(child);
        this.write(parent, jsonObject);

        this.checkRemoveEmpty();

        return true;
    }

    public boolean writeIfAbsent(final @NotNull String path, final @Nullable Object object) {
        if (this.has(path))
            return true;
        return write(path, object);
    }
    public boolean writeMany(final @NotNull Map<String, Object> map) {
        final AtomicBoolean success = new AtomicBoolean(true);
        map.forEach((string, object) -> {
            if (!this.write(string, object))
                success.set(false);
        });

        return success.get();
    }
    public boolean write(final @NotNull String path, final @Nullable Object object) {
        return this.write(path, asJsonObject(), object) != null;
    }
    private JSONObject write(final @NotNull String path, final @NotNull JSONObject jsonObject, @Nullable Object object) {
        if (object == null) {
            this.remove(path);
            return jsonObject;
        }

        final ArrayList<String> parts = new ArrayList<>(List.of(path.split("\\.")));
        final JSONObject original = asJsonObject();
        JSONObject current = original;

        for (int i = 0; i < parts.size(); i++) {
            final String part = parts.get(i);
            final @Nullable Object currentObject = current.get(part);

            if (i == parts.size() - 1) {
                if (object instanceof Enum<?> anEnum)
                    object = anEnum.toString();
                if (object instanceof JsonManager json)
                    object = json.asJsonObject();

                current.put(part, object);
                this.content = original.toJSONString();
                return current;
            }

            if (currentObject instanceof JSONObject jObject) {
                current = jObject;
                continue;
            }

            if (currentObject == null) {
                final JSONObject newObject = new JSONObject();
                current.put(part, newObject);
                current = newObject;
                continue;
            }
            return null;
        }
        return null;
    }

    private boolean checkRemoveEmpty() {
        return this.checkRemoveEmpty(asJsonObject(), new StringBuilder());
    }
    private boolean checkRemoveEmpty(final @NotNull JSONObject jsonObject, @NotNull StringBuilder path) {
        for (final Object object : jsonObject.keySet()) {
            if (!(object instanceof JSONObject newJsonObject))
                continue;
            if (!checkRemoveEmpty(newJsonObject, path.append(path.isEmpty() ? "" : ".").append(object)))
                continue;
            if (!newJsonObject.isEmpty())
                continue;
            jsonObject.remove(object);
        }
        return false;
    }

    public boolean has(String path) {
        return getObject(path) != null;
    }

    public boolean isOriginalContent() {
        return content.equals(originalContent);
    }
    public boolean isModifiedContent() {
        return !isOriginalContent();
    }

    public Document toDocument() {
        return Document.parse(content);
    }
    public BsonDocument toBsonDocument() {
        return toDocument().toBsonDocument();
    }
    public String toString() {
        return getContent();
    }
}

File: src/main/java/at/flauschigesalex/defaultLibrary/file/JsonManager.kt
@file:Suppress("MemberVisibilityCanBePrivate", "unused")

package at.flauschigesalex.defaultLibrary.file

import org.bson.BsonDocument
import org.bson.Document
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import java.io.File
import java.net.http.HttpResponse

class JsonManager private constructor(content: String) {

    companion object {

        operator fun invoke() : JsonManager {
            return JsonManager("{}")
        }

        operator fun invoke(jsonString: String?) : JsonManager? {
            try {
                JSONParser().parse(jsonString)
                return JsonManager(jsonString!!)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return null
        }
        operator fun invoke(jsonObject: Any?) : JsonManager? {
            return invoke(jsonObject.toString())
        }
        operator fun invoke(map: Map<String, Any?>) : JsonManager {
            return this().apply { this.writeMany(map.filter { it.value != null }) }
        }
        operator fun invoke(vararg pairs: Pair<String, Any?>) : JsonManager {
            return this(mapOf(*pairs))
        }
        operator fun invoke(pairs: Collection<Pair<String, Any?>>) : JsonManager {
            return this(pairs.toMap())
        }
        operator fun invoke(data: DataHandler?) : JsonManager? {
            if (data == null)
                return null
            
            return invoke(data.readString())
        }
        operator fun invoke(file: File) : JsonManager? {
            return this(FileHandler(file))
        }
        operator fun invoke(document: Document) : JsonManager? {
            return this(document.toJson())
        }
    }

    var originalContent: String
        private set
    var content: String
        private set

    init {
        try {
            JSONParser().parse(content)
        } catch (e: Exception) {
            throw IllegalArgumentException("Failed to read json from string:\n$content")
        }

        this.originalContent = content
        this.content = content
    }

    private fun get(jsonObject: JSONObject, path: String): Any? {
        if (!jsonObject.containsKey(path))
            return null

        return jsonObject[path]
    }

    fun getObject(path: String): Any? {
        if (path.isEmpty()) return toObject()

        val jsonObject = toJsonObject()

        if (!path.contains("."))
            return get(jsonObject, path)

        val completed = arrayListOf<String>()
        var currentJO = jsonObject

        val splitSourcePath = path.split(".").toList()
        for (splitSource in splitSourcePath) {
            val obj = get(currentJO, splitSource)
            val currentPath = completed.joinToString(separator = ".")+"."+splitSource

            if (currentPath == path)
                return obj ?: if (path.endsWith("._")) null else this.getObject("$path._")

            try {
                completed.add(splitSource)
                currentJO = obj as? JSONObject ?: return null
            } catch (fail: Exception) {
                fail.printStackTrace()
                break
            }
        }
        return null
    }

    fun getJsonObject(path: String): JSONObject? {
        if (path.isEmpty())
            return toJsonObject()

        try {
            return getObject(path) as JSONObject?
        } catch (ignore: Exception) {}
        return null
    }

    fun values(path: String = ""): Collection<Any> {
        val jsonObject = getJsonObject(path) ?: return listOf()
        return jsonObject.values.filterNotNull()
    }

    fun random(path: String = ""): Any {
        return this.randomOrNull(path)!!
    }
    fun randomOrNull(path: String = ""): Any? {
        return values(path).randomOrNull()
    }

    fun getJsonManager(path: String): JsonManager? {
        try {
            return JsonManager(getJsonObject(path)!!)
        } catch (ignore: Exception) {}
        return null
    }

    fun getObjectList(path: String): List<Any> {
        try {
            return getObject(path).let { it as List<*> }.map { it as Any }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getJsonList(path: String): List<JsonManager> {
        try {
            return getStringList(path).map { JsonManager(it) }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getStringList(path: String): List<String> {
        try {
            return getObjectList(path).map { it.toString() }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getIntegerList(path: String): List<Int> {
        try {
            return getStringList(path).map { it.toInt() }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getLongList(path: String): List<Long> {
        try {
            return getStringList(path).map { it.toLong() }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getShortList(path: String): List<Short> {
        try {
            return getStringList(path).map { it.toShort() }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getDoubleList(path: String): List<Double> {
        try {
            return getStringList(path).map { it.toDouble() }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getFloatList(path: String): List<Float> {
        try {
            return getStringList(path).map { it.toFloat() }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getBooleanList(path: String): List<Boolean> {
        try {
            return getObjectList(path).map { it as Boolean }
        } catch (ignore: Exception) {}

        return listOf()
    }

    fun getString(path: String): String? {
        val obj = getObject(path) ?: return null
        return obj.toString()
    }

    fun getInteger(path: String): Int? {
        try {
            return getString(path)!!.toInt()
        } catch (ignore: Exception) {}
        return null
    }

    fun getLong(path: String): Long? {
        try {
            return getString(path)!!.toLong()
        } catch (ignore: Exception) {}
        return null
    }

    fun getShort(path: String): Short? {
        try {
            return getString(path)!!.toShort()
        } catch (ignore: Exception) {}
        return null
    }

    fun getDouble(path: String): Double? {
        try {
            return getString(path)!!.toDouble()
        } catch (ignore: Exception) {}
        return null
    }

    fun getFloat(path: String): Float? {
        try {
            return getString(path)!!.toFloat()
        } catch (ignore: Exception) {}
        return null
    }

    fun getBoolean(path: String): Boolean? {
        try {
            return getString(path).toBoolean()
        } catch (ignore: Exception) {}
        return null
    }

    @Deprecated("", level = DeprecationLevel.ERROR)
    fun copyTo(path: String, newPath: String, override: Boolean = true): Boolean {
        return false
    }

    @Deprecated("", level = DeprecationLevel.ERROR)
    fun move(path: String, newPath: String, override: Boolean = true): Boolean {
        return false
    }

    fun removeMany(vararg paths: String) {
        return this.removeMany(paths.toList())
    }

    fun removeMany(paths: Collection<String>) {
        for (path in paths)
            this.remove(path)
    }

    fun remove(path: String) {
        if (!this.contains(path))
            return

        if (!path.contains(".")) {
            val jsonObject = this.toJsonObject()
            jsonObject.remove(path)
            this.content = jsonObject.toString()

            this.checkRemoveEmpty()
            return
        }

        val child: String = path.split(".").last()
        val parent = path.substring(0, path.length - (child.length + 1))

        val jsonObject = this.getJsonObject(parent)
        jsonObject!!.remove(child)
        this.write(parent, jsonObject)

        this.checkRemoveEmpty()

        return
    }

    fun writeIfAbsent(path: String, obj: Any?): Boolean {
        if (this.contains(path)) return true
        return write(path, obj)
    }

    fun writeMany(map: Map<String, Any?>): Boolean {
        return map.map { (path, obj) -> this.write(path, obj) }.all { it }
    }

    fun write(path: String, obj: Any?): Boolean {
        return this.write(path, toJsonObject(), obj) != null
    }

    private fun write(path: String, jsonObject: JSONObject, obj: Any?): JSONObject? {
        var temp = obj
        if (temp == null) {
            this.remove(path)
            return jsonObject
        }

        val parts = path.split(".").toList()
        val original = toJsonObject()
        var current = original

        for (i in parts.indices) {
            val part = parts[i]
            val currentObject = current[part]

            if (i == parts.size - 1) {
                if (temp is Enum<*>) temp = temp.toString()
                if (temp is JsonManager) temp = temp.toJsonObject()

                current[part] = temp
                this.content = original.toJSONString()
                return current
            }

            if (currentObject is JSONObject) {
                current = currentObject
                continue
            }

            if (currentObject == null) {
                val newObject = JSONObject()
                current[part] = newObject
                current = newObject
                continue
            }
            return null
        }
        return null
    }

    private fun checkRemoveEmpty(
        jsonObject: JSONObject = toJsonObject(),
        path: StringBuilder = StringBuilder()
    ): Boolean {
        for (obj in jsonObject.keys) {
            if (obj !is JSONObject)
                continue

            if (!checkRemoveEmpty(obj, path.append(if (path.isEmpty()) "" else ".").append(obj)))
                continue

            if (obj.isNotEmpty())
                continue

            jsonObject.remove(obj)
        }
        return false
    }

    fun contains(path: String): Boolean {
        return getObject(path) != null
    }

    fun isOriginalContent(): Boolean {
        return content == originalContent
    }
    fun isModifiedContent(): Boolean {
        return !isOriginalContent()
    }
    
    fun toObject(): Any {
        return JSONParser().parse(content)
    }
    fun toJsonObject(): JSONObject {
        return toObject() as JSONObject
    }
    fun toDocument(): Document {
        return Document.parse(content)
    }
    fun toBsonDocument(): BsonDocument {
        return toDocument().toBsonDocument()
    }
    override fun toString(): String {
        return content
    }

    fun copy(): JsonManager {
        val json = JsonManager(content)
        json.originalContent = originalContent

        return json
    }
    fun copyOriginal(): JsonManager {
        return JsonManager(originalContent)
    }
}
